File: backend/backend_Crypto_Purchase_Endpoint/src/services/purchaseService.js
const { withTransaction } = require("../db");

class PurchaseError extends Error {
  constructor(message, statusCode) {
    super(message);
    this.name = "PurchaseError";
    this.statusCode = statusCode;
  }
}

function normalizeSymbol(symbol) {
  return String(symbol || "").trim().toUpperCase();
}

function validatePurchaseInput({ userId, symbol, amount }) {
  if (!Number.isInteger(userId) || userId <= 0) {
    throw new PurchaseError("userId must be a positive integer", 400);
  }

  if (typeof symbol !== "string" || symbol.trim().length === 0 || symbol.trim().length > 10) {
    throw new PurchaseError("symbol must be a non-empty string up to 10 chars", 400);
  }

  if (typeof amount !== "number" || amount <= 0) {
    throw new PurchaseError("amount must be a positive number", 400);
  }
}

function buildPurchaseService(deps = {}) {
  const runWithTransaction = deps.withTransaction || withTransaction;

  return {
    async purchaseCoin({ userId, symbol, amount }) {
      validatePurchaseInput({ userId, symbol, amount });
      const normalizedSymbol = normalizeSymbol(symbol);

      return runWithTransaction(async (connection) => {
        const [users] = await connection.execute(
          "SELECT id FROM users WHERE id = ?",
          [userId]
        );
        if (users.length === 0) {
          throw new PurchaseError("User not found", 404);
        }

        const [coins] = await connection.execute(
          "SELECT symbol, current_price_usd, circulating_supply FROM coins WHERE symbol = ? FOR UPDATE",
          [normalizedSymbol]
        );
        if (coins.length === 0) {
          throw new PurchaseError("Coin not found", 404);
        }

        const coin = coins[0];
        const circulatingSupply = Number(coin.circulating_supply);
        const coinPriceUsd = Number(coin.current_price_usd);

        if (circulatingSupply < amount) {
          throw new PurchaseError("Insufficient circulating supply", 409);
        }

        await connection.execute(
          "UPDATE coins SET circulating_supply = circulating_supply - ? WHERE symbol = ?",
          [amount, normalizedSymbol]
        );

        await connection.execute(
          `INSERT INTO wallets (user_id, symbol, balance)
          VALUES (?, ?, ?)
          ON DUPLICATE KEY UPDATE balance = balance + VALUES(balance)`,
          [userId, normalizedSymbol, amount]
        );

        const [walletRows] = await connection.execute(
          "SELECT user_id AS userId, symbol, balance FROM wallets WHERE user_id = ? AND symbol = ?",
          [userId, normalizedSymbol]
        );

        const totalUsd = Number((amount * coinPriceUsd).toFixed(2));
        const [insertResult] = await connection.execute(
          `INSERT INTO transactions
          (user_id, symbol, transaction_type, quantity, price_usd, total_usd)
          VALUES (?, ?, 'BUY', ?, ?, ?)`,
          [userId, normalizedSymbol, amount, coinPriceUsd, totalUsd]
        );

        const [transactionRows] = await connection.execute(
          `SELECT id, user_id AS userId, symbol, transaction_type AS transactionType,
          quantity, price_usd AS priceUsd, total_usd AS totalUsd, created_at AS createdAt
          FROM transactions
          WHERE id = ?`,
          [insertResult.insertId]
        );

        return {
          message: "Coin purchase successful",
          transaction: transactionRows[0],
          wallet: walletRows[0]
        };
      });
    }
  };
}

module.exports = {
  PurchaseError,
  buildPurchaseService
};

File: backend/backend_Crypto_Purchase_Endpoint/src/main/java/com/cs506/cryptopurchase/service/PurchaseService.java
package com.cs506.cryptopurchase.service;

import com.cs506.cryptopurchase.dto.PurchaseResponse;
import com.cs506.cryptopurchase.dto.TransactionDto;
import com.cs506.cryptopurchase.dto.WalletDto;
import com.cs506.cryptopurchase.exception.PurchaseException;
import com.cs506.cryptopurchase.model.CoinSnapshot;
import com.cs506.cryptopurchase.model.TransactionSnapshot;
import com.cs506.cryptopurchase.model.WalletSnapshot;
import com.cs506.cryptopurchase.repository.PurchaseRepository;
import java.math.BigDecimal;
import java.math.RoundingMode;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PurchaseService {

    private final PurchaseRepository purchaseRepository;

    public PurchaseService(PurchaseRepository purchaseRepository) {
        this.purchaseRepository = purchaseRepository;
    }

    @Transactional
    public PurchaseResponse purchaseCoin(Integer userId, String symbol, BigDecimal amount) {
        validateInput(userId, symbol, amount);
        String normalizedSymbol = symbol.trim().toUpperCase();

        if (!purchaseRepository.userExists(userId)) {
            throw new PurchaseException("User not found", 404);
        }

        CoinSnapshot coin = purchaseRepository.lockCoin(normalizedSymbol)
            .orElseThrow(() -> new PurchaseException("Coin not found", 404));

        if (coin.circulatingSupply().compareTo(amount) < 0) {
            throw new PurchaseException("Insufficient circulating supply", 409);
        }

        purchaseRepository.decrementSupply(normalizedSymbol, amount);
        purchaseRepository.upsertWallet(userId, normalizedSymbol, amount);

        WalletSnapshot wallet = purchaseRepository.findWallet(userId, normalizedSymbol)
            .orElseThrow(() -> new PurchaseException("Wallet not found after purchase", 500));

        BigDecimal totalUsd = coin.currentPriceUsd()
            .multiply(amount)
            .setScale(2, RoundingMode.HALF_UP);

        long transactionId = purchaseRepository.insertTransaction(
            userId,
            normalizedSymbol,
            amount,
            coin.currentPriceUsd(),
            totalUsd
        );

        TransactionSnapshot transaction = purchaseRepository.findTransaction(transactionId)
            .orElseThrow(() -> new PurchaseException("Transaction not found after insert", 500));

        return new PurchaseResponse(
            "Coin purchase successful",
            new TransactionDto(
                transaction.id(),
                transaction.userId(),
                transaction.symbol(),
                transaction.transactionType(),
                transaction.quantity(),
                transaction.priceUsd(),
                transaction.totalUsd(),
                transaction.createdAt()
            ),
            new WalletDto(
                wallet.userId(),
                wallet.symbol(),
                wallet.balance()
            )
        );
    }

    private void validateInput(Integer userId, String symbol, BigDecimal amount) {
        if (userId == null || userId <= 0) {
            throw new PurchaseException("userId must be a positive integer", 400);
        }

        if (symbol == null || symbol.trim().isEmpty() || symbol.trim().length() > 10) {
            throw new PurchaseException("symbol must be a non-empty string up to 10 chars", 400);
        }

        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            throw new PurchaseException("amount must be a positive number", 400);
        }
    }
}
